src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -0,0 +1,227 @@
+import org.opensearch.common.xcontent.XContentParser;
+import org.opensearch.index.mapper.NumberFieldMapper;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.QueryShardContext;
+
+/**
+ * NeuralQueryBuilder is responsible for producing "neural" query types. A "neural" query type is a wrapper around a
+ * k-NN vector query. It uses a ML language model to produce a dense vector from a query string that is then used as
+ * the query vector for the k-NN search.
+ */
+
+@Log4j2
+@Getter
+@NoArgsConstructor
+public class NeuralQueryBuilder extends AbstractQueryBuilder<NeuralQueryBuilder> {
+
+    public static final String NAME = "neural";
+
+    static final ParseField QUERY_TEXT_FIELD = new ParseField("query_text");
+
+    static final ParseField MODEL_ID_FIELD = new ParseField("model_id");
+
+    static final ParseField K_FIELD = new ParseField("k");
+
+    private String fieldName;
+    private String queryText;
+    private String modelId;
+    private int k;
+
+    /**
+     * Set the fieldName this query will be executed against
+     *
+     * @param fieldName name of k-NN vector field that query will be executed against
+     * @return this
+     */
+    public NeuralQueryBuilder fieldName(String fieldName) {
+        this.fieldName = fieldName;
+        return this;
+    }
+
+    /**
+     * Set the queryText that will be translated into the dense query vector used for k-NN search.
+     *
+     * @param queryText Text of a query that should be translated to a dense vector
+     * @return this
+     */
+    public NeuralQueryBuilder queryText(String queryText) {
+        this.queryText = queryText;
+        return this;
+    }
+
+    /**
+     * Set the modelId that should produce the dense query vector
+     *
+     * @param modelId ID of model to produce query vector
+     * @return this
+     */
+    public NeuralQueryBuilder modelId(String modelId) {
+        this.modelId = modelId;
+        return this;
+    }
+
+    /**
+     * Set the number of neighbors that should be retrieved during k-NN search
+     *
+     * @param k number of neighbors to be retrieved in k-NN query
+     * @return this
+     */
+    public NeuralQueryBuilder k(int k) {
+        this.k = k;
+        return this;
+    }
+
+    /**
+     * Constructor from stream input
+     *
+     * @param in StreamInput to initialize object from
+     * @throws IOException thrown if unable to read from input stream
+     */
+    public NeuralQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.queryText = in.readString();
+        this.modelId = in.readString();
+        this.k = in.readInt();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(this.fieldName);
+        out.writeString(this.queryText);
+        out.writeString(this.modelId);
+        out.writeInt(this.k);
```

Comment:
see if want to use writeVInt() instead of Int. VInt uses less bytes than int.

Reply:
Will update